File: backend/src/main/java/com/tmk/vtcmanager/application/domain/operation/OperationFinanciere.java
package com.tmk.vtcmanager.application.domain.operation;

import com.tmk.vtcmanager.application.domain.chauffeur.Chauffeur;
import com.tmk.vtcmanager.application.domain.partenaire.Partenaire;
import com.tmk.vtcmanager.application.domain.vehicule.Vehicule;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.Set;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class OperationFinanciere {

    private Long id;
    private String reference;
    private TypeOperation typeOperation;
    private CategorieOperation categorie;
    private SousCategorieOperation sousCategorie;
    private Chauffeur chauffeur;
    private Vehicule vehicule;
    /**
     * Tiers de l'écriture — le garage payé comptant, l'assureur, le bailleur.
     * Facultatif : toutes les écritures n'ont pas de contrepartie externe.
     */
    private Partenaire partenaire;
    private BigDecimal montant;
    private ModePaiement modePaiement;
    /**
     * Compte de trésorerie mouvementé. Null toléré (opérations legacy sans
     * compte) : résolu au compte par défaut du type dérivé du mode de
     * paiement lors de la création.
     */
    private Long compteTresorerieId;
    private LocalDate dateOperation;
    /**
     * Date "métier" de référence : pour un encaissement de période, c'est la date
     * de la période concernée (recette/cotisation/faute) et non la date de la
     * transaction. Sert à l'affichage des lignes d'opération. Null = utiliser
     * {@link #dateOperation}.
     */
    private LocalDate dateReference;
    private String commentaire;
    private StatutOperation statut;
    private DetailMaintenance detailMaintenance;

    /**
     * Maintenance à l'origine de cette opération (dépense générée par la
     * complétion d'une maintenance). Null pour les autres opérations. Sert à
     * rouvrir la maintenance lorsqu'on annule l'opération.
     */
    private Long maintenanceId;

    /**
     * Contravention réglée par cette écriture (remboursement encaissé auprès du
     * chauffeur, ou compensation lors d'un arrêté de compte). Null pour les
     * autres opérations. Sert à repositionner la contravention lorsqu'on annule
     * l'opération.
     */
    private Long contraventionId;

    /**
     * Écriture contre-passée par celle-ci. Renseigné sur l'extourne seule, dont
     * le montant est l'opposé de l'origine — le couple s'annule dans tous les
     * agrégats sans qu'aucune écriture ne disparaisse.
     */
    private Long extourneDeId;

    /**
     * Renseigné à la lecture seulement : {@link #estAnnulable()} — ce que l'état
     * de l'écriture permet — combiné au verrou des arrêtés, que seule la couche
     * de lecture connaît. Faux, le client masque l'action « Annuler ».
     */
    private Boolean annulable;

    /** Motif d'annulation, porté par l'écriture d'origine. */
    private String motifAnnulation;
    private String annulePar;
    private LocalDateTime annuleLe;

    /**
     * Facture fournisseur soldée par cette écriture. Renseigné sur un règlement :
     * la charge étant déjà portée par la facture, la lecture en base engagement
     * écarte ces écritures pour ne pas la compter deux fois.
     */
    private Long facturePartenaireId;

    /** Vrai si cette écriture a été contre-passée. */
    public boolean estExtournee() {
        return annuleLe != null;
    }

    /** Vrai si cette écriture est elle-même une contre-passation. */
    public boolean estUneExtourne() {
        return extourneDeId != null;
    }

    /**
     * Catégories dont l'écriture solde une créance tenue ailleurs : ligne de
     * recette, de cotisation, de pénalité, ou contravention. Leur montant et
     * leur rattachement ne s'éditent pas en place — ils sont le reflet d'un
     * versement enregistré dans le module d'origine.
     */
    private static final Set<String> CODES_ENCAISSEMENT = Set.of(
            "ENCAISSEMENT_RECETTES",
            "ENCAISSEMENT_COTISATIONS",
            "ENCAISSEMENT_PENALITES",
            "CONTRAVENTION_REMBOURSEMENT");

    /** Vrai si l'écriture règle une créance suivie dans un autre module. */
    public boolean estUnEncaissement() {
        return categorie != null && categorie.getCode() != null
                && CODES_ENCAISSEMENT.contains(categorie.getCode().toUpperCase());
    }

    /** Vrai si l'écriture est la dépense née de la complétion d'une maintenance. */
    public boolean estIssueDUneMaintenance() {
        return maintenanceId != null;
    }

    /**
     * Vrai si l'écriture peut encore être retouchée en place.
     *
     * <p>Un encaissement et une dépense de maintenance en sont exclus : leur
     * montant appartient à la ligne ou à l'intervention qui les a produits, et
     * le corriger ici les désaccorderait en silence. La voie est l'annulation,
     * qui repositionne la source, puis la ressaisie.
     */
    public boolean isModifiable() {
        return statut != StatutOperation.ANNULEE
                && !estUneExtourne()
                && !estExtournee()
                && !estUnEncaissement()
                && !estIssueDUneMaintenance();
    }

    /**
     * Vrai si l'écriture peut encore être contre-passée, au seul vu de son état.
     *
     * <p>Trois écritures ne s'annulent pas, et pour la même raison : la
     * correction a déjà eu lieu. Une extourne <em>est</em> la correction —
     * l'extourner rendrait vie à l'origine par la bande. Une écriture extournée
     * l'a déjà reçue — une seconde contre-passation la compterait deux fois en
     * négatif. Une écriture passée {@code ANNULEE} sans extourne — l'annulation
     * d'un arrêté de compte, par exemple — a été neutralisée ailleurs.
     *
     * <p>Le verrou des arrêtés ne se lit pas ici : il porte sur la date, pas sur
     * l'écriture, et se combine à ce jugement dans le drapeau {@code annulable}
     * envoyé au client. La règle métier, elle, tient à cet endroit — les clients
     * n'ont plus à la rejouer chacun de leur côté.
     */
    public boolean estAnnulable() {
        return statut != StatutOperation.ANNULEE
                && !estUneExtourne()
                && !estExtournee();
    }
}

File: mobile/lib/features/operation_financiere/domain/entities/operation_financiere.dart
import '../../../../core/utils/libelle_operation.dart' as date_relative;
import 'detail_maintenance.dart';
import '../enums/mode_paiement.dart';
import '../enums/statut_operation.dart';
import '../enums/type_operation.dart';

class OperationFinanciere {
  final int? id;
  final String? reference;
  final TypeOperation typeOperation;
  final int? categorieId;
  final String? categorieCode;
  final String? categorieLibelle;
  final int? sousCategorieId;
  final String? sousCategorieLibelle;
  final String? sousCategorieCode;
  final int? chauffeurId;
  final String? chauffeurNom;
  final int? vehiculeId;
  final String? vehiculeNom;

  /// Tiers de l'écriture : garage, assureur, bailleur… Facultatif.
  final int? partenaireId;
  final String? partenaireNom;
  final double montant;
  final ModePaiement? modePaiement;
  final DateTime dateOperation;

  /// Date "métier" de référence (date de la période concernée pour un
  /// encaissement : recette / cotisation / faute). Null pour les autres
  /// opérations → on retombe sur [dateOperation].
  final DateTime? dateReference;
  final String? commentaire;
  final StatutOperation statut;
  final DetailMaintenance? detailMaintenance;

  /// Écriture contre-passée par celle-ci : non nul sur une extourne, dont le
  /// montant est l'opposé de l'origine.
  final int? extourneDeId;

  /// Renseignés sur une écriture qui a été extournée.
  final DateTime? annuleLe;
  final String? motifAnnulation;

  /// Faux pour une écriture que le backend refuse de retoucher en place :
  /// encaissement (recette, cotisation, pénalité, contravention), dépense issue
  /// d'une maintenance, extourne ou écriture déjà extournée. L'action
  /// « Modifier » est alors masquée ; l'annulation, elle, reste ouverte.
  final bool modifiable;

  /// Faux quand le backend refuserait la contre-passation : écriture déjà
  /// corrigée (extourne, écriture extournée ou annulée), ou arrêté — période
  /// comptable close, caisse comptée — couvrant sa date. Le bouton « Annuler »
  /// est alors masqué. La règle est tenue côté serveur : ne pas la redéduire
  /// ici, les deux versions divergeraient.
  ///
  /// Faux par défaut quand la réponse ne porte pas le drapeau : en l'absence
  /// d'information, on ne propose pas une action qui pourrait être refusée.
  final bool annulable;

  /// Cette écriture est une contre-passation.
  bool get estUneExtourne => extourneDeId != null;

  /// Cette écriture a été contre-passée : elle reste au journal, neutralisée.
  bool get estExtournee => annuleLe != null;

  const OperationFinanciere({
    this.id,
    this.reference,
    required this.typeOperation,
    this.categorieId,
    this.categorieCode,
    this.categorieLibelle,
    this.sousCategorieId,
    this.sousCategorieLibelle,
    this.sousCategorieCode,
    this.chauffeurId,
    this.chauffeurNom,
    this.vehiculeId,
    this.vehiculeNom,
    this.partenaireId,
    this.partenaireNom,
    required this.montant,
    this.modePaiement,
    required this.dateOperation,
    this.dateReference,
    this.commentaire,
    this.statut = StatutOperation.ENCAISSE,
    this.detailMaintenance,
    this.extourneDeId,
    this.annuleLe,
    this.motifAnnulation,
    this.modifiable = false,
    this.annulable = false,
  });

  /// Date à afficher sur les lignes d'opération : la date métier si présente
  /// (encaissement de période), sinon la date de l'opération.
  DateTime get dateAffichee => dateReference ?? dateOperation;

  /// Connecteur de date relatif de la date métier, ex. « d'hier ».
  /// Voir [libelleDateRelative] (partagé avec le rapport financier).
  String get libelleDateRelative => date_relative.libelleDateRelative(dateAffichee);

  /// Vrai si l'opération appartient au groupe "Maintenances"
  /// (déterminé par le libellé de la sous-catégorie côté backend,
  ///  sans dépendre d'un code de catégorie figé).
  bool get isMaintenance =>
      sousCategorieLibelle?.toLowerCase() == 'maintenances';

  /// Vrai si l'opération est un encaissement (recette / cotisation / pénalité).
  /// Seuls ces types conservent la date relative dans leur libellé de ligne ;
  /// les autres opérations (dépenses, maintenance…) s'affichent sans date.
  bool get estEncaissement =>
      date_relative.estCategorieEncaissement(categorieCode);

  /// Libellé du titre affiché sur les lignes d'opération (Accueil, liste des
  /// opérations, rapport financier) : « Catégorie d'hier » pour un
  /// encaissement, « Catégorie » seul sinon.
  String get libelleLigne => date_relative.libelleLigneOperation(
        categorieCode: categorieCode,
        categorie: categorieLibelle ?? typeOperation.libelle,
        date: dateAffichee,
      );
}
